packages/payment-detection/src/erc777/superfluid-retriever.ts

```diff
@@ -0,0 +1,90 @@
+import { PaymentTypes } from '@requestnetwork/types';
+import { utils } from 'ethers';
+import { FlowUpdatedEvent } from '../thegraph/generated/graphql-superfluid';
+import { getTheGraphClient, TheGraphClient, TheGraphClientOptions } from '../thegraph/superfluid';
+
+/** Parameters for getting payment events from theGraph */
+type GraphPaymentQueryParams = {
+  reference: string;
+  to: string;
+  tokenAddress: string | null;
+};
+
+export class SuperFluidInfoRetriever {
+  private client: TheGraphClient;
+
+  /**
+   * @param paymentReference The reference to identify the payment
+   * @param tokenContractAddress The address of the ERC777 contract
+   * @param toAddress Address of the balance we want to check
+   * @param eventName Indicate if it is an address for payment or refund
+   * @param network The Ethereum network to use
+   * @param options Extra options to GraphQL client
+   */
+  constructor(
+    private paymentReference: string,
+    private tokenContractAddress: string | null,
+    private toAddress: string,
+    private eventName: PaymentTypes.EVENTS_NAMES,
+    private network: string,
+    private options?: TheGraphClientOptions,
+  ) {
+    this.client = getTheGraphClient(this.network, this.options);
+  }
+
+  private getGraphVariables(): GraphPaymentQueryParams {
+    return {
+      reference: utils.keccak256(`0x${this.paymentReference}`),
+      to: this.toAddress,
+      tokenAddress: this.tokenContractAddress,
+    };
+  }
+
+  public async getTransferEvents(): Promise<PaymentTypes.ERC20PaymentNetworkEvent[]> {
```

Comment:
this method deserves a lot of comments! the logic isn't trivial. 

Besides, it's not robust enough I think. you need to consider the possibility of not finding any matching events... here i'm pretty sure it would throw an error

Reply:
descripion done

Reply:
I've made the function more robust by returning [] when no matching are found.
I've commented, simplified and secured the computation of amounts to only consider needed pair of events.